### DISCUSSION TITLE
LNM - OpenLayers Integration Roadmap

### DISCUSSION THREAD

**Discussion by KOKAProduktion** [albar965/littlenavmap#677] [Ideas]
Hiho @albar965, hiho @u-an-i,

i finally got some thoughts together after diving deeper into LNM's structure, its repos, your discussions and the [littlenavmap-openlayers implementation](https://github.com/KOKAProduktion/littlenavmap-openlayers)s current state. 

Wow, @u-an-i you're really tackling a lot! I admire your drive and ambition to modify/refactor and to introduce a lot of new ideas. I hope it's ok for you, if i present my current view mostly focused on the OpenLayers-related topics only and link to your proposals wherever they overlap. I'm sure we can manage to join forces to contribute something cool to @albar965's LNM project.

--------------------------------

# LNM - OpenLayers Integration Roadmap

At the time of writing LNM is serving the [Marble](https://marble.kde.org) rendered map image via the built in http-server for requesting web clients. The UI/UX logic for map interaction besides other features is a custom implementation.

[OpenLayers](https://github.com/openlayers/openlayers) (OL) is a web-based mapping framework written in JS under the [BSD-2-Clause License](https://github.com/openlayers/openlayers/blob/main/LICENSE.md) providing the following features i.a.:

- Retrieval and rendering 
    - of both tiled and untiled map sources organized on layers
    - of vector data
    - of stylable map features like clickable items
- Popular map navigation UX (drag to pan, mouse zoom, etc.)
- API to manipulate the maps behaviour programmatically (e.g. animate to position, etc.)

The following comments show a possible roadmap based on my review and assumptions. Please feel free to correct me and to add your thoughts on the topics!

**Comment by KOKAProduktion** [comment]
## Milestone 1 : LNM Web API

The idea is to expose dynamic elements information (e.g. position and appearance of Airports, VOR's, the users aircraft etc.) which is currently being composed onto the LNM Marble image output, via a Web API.

### Requirements

- The ability to retrieve the required LNM data in the built-in webserver context
- Extending LNM by a web-based API to query interactable map features (e.g. Airports, VOR's, etc.), the users aircraft (e.g. position, orientation, vector flight path etc.) and "output image static" elements (e.g. Contributions, Wind indicator)

### Thoughts

Creating a strict API contract allows full separation of concerns between LNM and possible web UI implementations. Both sides could develop independently as long as they satisfy the contract.

### Related

- #663
- #150
- #397

### Possible Solution

- Defining a sensible API contract using
    - [OpenAPI](https://oai.github.io/Documentation/)
- Providing the build in webserver context with the required data
- Extending the built-in webserver by a JSON/RESTful API satisfying the contract mentioned above using
    - [JSON Support in Qt](https://doc.qt.io/qt-5/json.html)
    - [nlohmann/json](https://github.com/nlohmann/json) as proposed by @u-an-i 

### PR

- #776

**Reply by KOKAProduktion** [reply]
Created a LNM Web API proposal PoC at: https://github.com/albar965/littlenavmap/pull/731

**Reply by KOKAProduktion** [reply]
Extended API proposal by

- simulation aircraft info
- UI info

at: #772

**Reply by KOKAProduktion** [reply]
Minor fixes at: https://github.com/albar965/littlenavmap/pull/776

**Comment by KOKAProduktion** [comment]
## Milestone 2 : LNM WMTS API (optional)

The idea is to modify the LNM Marble image output web routes to satisfy a strict WMTS API definition in order to increase compatibility with mapping clients like OL. This could come in handy if OL should become abandoned/obsolete and would require replacement by another WMTS-capable mapping solution.

However this should be considered optional as its map image retrieval would become obsolete with milestone 5.

### Possible Solution

- Modifying LNM Marble image and data web retrieval routes to reflect a standard defined API specification
    - [OpenGIS Web Map Tile Service Implementation Standard](https://www.ogc.org/standards/wmts)

**Comment by KOKAProduktion** [comment]
## Milestone 3 : LNM Web UI OL Display Integration

The idea is to replace the existing **web ui map display** of the rendered Marble image from LNM by an OL implementation to remove the need for custom mapping and map navigation logic and to provide a popular map UX. LNM remains serving the Marble images as the map source but requested in multiple tiles like a WMTS server instead of a single image.

Proof of Concept: [KOKAProduktion/littlenavmap-openlayers](https://github.com/KOKAProduktion/littlenavmap-openlayers) first presented at #622

### Requirements

- LNM Web API developed at milestone 1
- (optional) LNM WMTS API developed at milestone 2

### Related

- #665

### Possible Solution

- Modification of the LNM Marble image output rendering for tile-based display
    - Hiding of common/static elements (e.g. Contributions, Wind indicator)
    - Hiding of interactable/dynamic elements (e.g. Airports, VOR's, the users aircraft etc.)
- The actual OL implementation assembling the data retrieved from the Web API visually to enable interactability with the map features and to avoid rendering issues with already composed images.

### PR

- #826

**Reply by KOKAProduktion** [reply]
Created image retrieval proposal at #788

**Reply by KOKAProduktion** [reply]
Hiho @albar965, hiho @u-an-i,

working on the OL integration has risen some questions for me about on how the projects should integrate meaningfully in terms of structuring. I'd be happy to hear your thoughts on the following:

- https://github.com/KOKAProduktion/littlenavmap-openlayers is a project which I'd like to limit on map-related features only. This should allow integrating it into any JS-driven web UI. Meaning it should scope on:
    - Map display and map navigation
    - Dynamic objects display (e.g. Aircraft)
    - Clickable features firing subscribeable events
        - #790 
    - Exposition of the OL API for map customization
- Since it's built in a **NodeJS** environment i think its sources should be kept separated from the Qt/C++ project to avoid increasing overall building complexity. I would suggest adding only the projects build/dist output to the LNM repo (`/web`) as already done in #788
- At a later point I'd like to start my own approach on a full LNM web UI focused on VR-compatibility integrating the OL implementation and using React.

What do you think?

Thanks for your time and good night!

**Reply by u-an-i** [reply]
hi @KOKAProduktion , welcome to October 8th :)


> * Map display and map navigation
> * Dynamic objects display (e.g. Aircraft)
> * Clickable features firing subscribeable events
> * Exposition of the OL API for map customization

i was thinking having the map without anything drawn upon and the aircraft as a separate image in the web ui, superimposed on the map (will have no JPEG artifacts of aircraft image already stitched in map image, allows having the aircraft fly around a stationary map display (aircraft not center) without requesting map images).
This could be applied to all thinkable map objects but might be quite some effort already done in main app and they are not moving relative to the map (at least airports don't). Otoh separate images for airports etc and map allow using native DOM features such as hover to create surplus values to obejcts.

An OL driven map display and interaction with it is one of the aims, this milestone. I would keep and would love to see it be as dependency-less and thus extra-installations-less, extra-documentation-reading-less and most likely (unfortunately from most of deps i know) very-very-much-maintainance-less as possible.

Adding to an OL driven map display and interaction would only be a UI to command such interactions and a way to automise/program such interaction or interaction targets, thus "toolbars" and "plugins" as far as in think of now. I attempted to provide toolbars and plugins already (toolbar offering yet can be much extended).

Thus you would only need to create a toolbar "system" and plugin "system" for users/developers of LNM to build upon. I assume a purpose-built (and) stream-lined system form scratch is the most flexible, lightweight and happyness-inducing solution possible.

Edited: corrected milestone reference

**Reply by u-an-i** [reply]
(ah yes, and the web api of course is a part (which can be utilised in plugins)

**Reply by KOKAProduktion** [reply]
Hiho @u-an-i,

Damn, you're quick! :) Thanks for your input!

> i was thinking having the map without anything drawn upon and the aircraft as a separate image in the web ui, superimposed on the map [...] allows having the aircraft fly around a stationary map display without requesting map images.

This is already working in #788 

> This could be applied to all thinkable map objects but might be quite some effort already done in main app [...] separate images for airports etc and map allow using native DOM features such as hover to create surplus values to obejcts

Yes, you're right. As a workaround i would keep the feature visuals on the map image but overlay them with invisible clickables to trigger feature related actions. I think as long as Marble is in use, this would be the fastest way to add interactability without recreating the features in HTML.

> I would keep and would love to see it be as dependency-less and thus extra-installations-less, extra-documentation-reading-less and most likely (unfortunately from most of deps i know) very-very-much-maintainance-less as possible

By including the OL implementation you have got the `map` object available in your JS scope and access to the documented OL API. No custom JS except the map, features and events retrieval here. This should also already work in #788 (well, except for the features ATTOW)

> Thus you would only need to create a toolbar "system" and plugin "system" for users/developers of LNM to build upon.

If you mean the map I think it's the other way around. Any JS-app can include the map and interact with it using the documented OL API or subscribe to its events. No need for the map to know anything beyond its mapping :)

If you're talking about the React approach, then yes, you're right. Some kind of predefined interfaces for plugins/extensions makes sense.

> I assume a purpose-built (and) stream-lined system form scratch is the most flexible, lightweight and happyness-inducing solution possible

Please forgive me, if i don't get this one right... I'd say especially because of maintenance-less and developer-friendly aspects I tend to use documented frameworks/solutions instead of reinventing the wheel from scratch... That's why I'd go for React, Angular or similar.

Again, thanks for your time and thoughts! They are very helpful!

**Reply by u-an-i** [reply]
> That's why I'd go for React, Angular or similar

If what they provide reliefs you of creating such... but they have no benefit whose achieving they are purpose-built for to achieve... needing Monstrous effort to apply for the first time... additional software on your pc, obscure storage locations, servers started on startup..  and the core of your to-be original (because if not original, why were you creating it?) creation isn't even touched yet to be understood extricated from the jaw of the jail they are in, ... and once it is, A has released 700 new version the readers past is useless

kis in every way

**Reply by KOKAProduktion** [reply]
Hmmmm... That's a bit of a question of ones philosophy, isn't it?

> but they have no benefit whose achieving they are purpose-built for to achieve... 

You mean that they are too generic and therefore have to be bended to our purpose? I see their benefit mostly in adding more advanced and, at least for me, desirable concepts to plain JS like:

- strict typing (TypeScript)
    - More "strictness" in JS makes the code more transient to the backend code language, enables "pre-compiling", consistency checks and sensible auto-complete. In my opinion it eases development and reduces errors.
- MVC (e.g. Angular)
    - I like MVC :)
    - Templating
- Live-data binding (e.g. Angular, React)
    - Fully managed UI "loop". No need for `setTimeouts` or "callback hell" anymore. 
- Dependency injection & JS-"compilation"/transpiling (webpack)
    - More control and optimisation options for the output package

I'd say they just extend the tools you have available for implementing your project. The less I have to implement, the better.

> needing Monstrous effort to apply for the first time... additional software on your pc, obscure storage locations, servers started on startup.. 

Can't really confirm that. My experience was extremely mind-blowing on what can be done with plain JS inside the node.js universe and how easy it is to apply. 

Angular and React can be challenging in the beginning but the reward is really f*** huge in my opinion! Of course I'm not claiming they are here to stay forever, but `XMLHttpRequest` usage got also less popular since `fetch` is natively available if you get my point.

> and the core of your to-be original (because if not original, why were you creating it?) 

I'm mostly writing applications for to be applied not to be original. If there is a stable and relatively established solution already available for a problem which is not specific to my project (like handling a map), I'll thankfully use it. The same way I'm building on top of LNM instead of writing my own fully integrated simulator mapping, weather, traffic, flight planning and monitoring app :)

> [...] from the jaw of the jail they are in

You can still keep your business logic in plain JS inside, if you're worried about those frameworks lifetimes.

> ...and once it is, A has released 700 new version the readers past is useless

Yes the update cycles are fast, but this also means free out-of-the-box updates for issues you don't want to worry about. Also, it's not that bad anymore as in node.js's early days.

> kis in every way

And that's probably the philosophical question here: Would you like to create a custom solution from scratch every time, or add/remove ready-made modules to/from an implementation which is orchestrating them. And where is the line if we use Qt with C++ and not just plain C for example?
In my experience custom implementations are never "simple", because they're... custom. Meaning everyone has to read-in into custom solutions everytime. Omitting the fact that they quite often lack documentation and the author has to maintain updates inside a changing environment.

In my opinion it's conventions and standards which make focusing on specific aspects of implementation work easier, leading to less specific code which translates to "simple" for me. At least I don't want to think about on how to fetch data or how to update a value in the UI every time. I can just look up and copy/paste the example from the corresponding frameworks docs...

If it's for the kicks I understand the urge to create individual solutions. But not for challenges/problems which have already been solved...

I hope you're not mad at me :) All paths lead to Rome...
If you haven't already tried out node.js, Angular and/or React please do! My guess would be, that you eventually could fall in love with at least TypeScript :) 

Thanks for your thoughts and cheers!

**Reply by u-an-i** [reply]
> > but they have no benefit whose achieving they are purpose-built for to achieve...
> 
> You mean that they are too generic and therefore have to be bended to our purpose?

No, they use concepts but those are not a solution to a problem, eg. MVC is just MVC (a naming), it's not pluggability like a DirectShow graph in which 1 part can be exchanged and everything still works right away, eg replace MySQL by PostgreSQL or replace view 1 by view 2. Read further on.

 
> * strict typing (TypeScript)
>   
>   * More "strictness" in JS makes the code more transient to the backend code language, enables "pre-compiling", consistency checks and sensible auto-complete. In my opinion it eases development and reduces errors.

TS is TS -- which caters for C++ developers (or perhaps Java and C# ones) -- and can be used separately, it is not a framework, throws away the strengths of JavaScript. Similar to JS 15++ (ECMA JS 6++): ECMA should have build upon the strengths of JS, instead they created ambiguity ("this"), unlegibility ("let" is l 1 | or I ?, => thin lines, is () now not a call anymore?), uneasy graspability (what does/should "let" mean?). Whether the advantages you list can/could not be achieved for plain JS I think not. "Luckily" you wrote "imo" because the effort to jump on TS and stay within certainly does not ease dev but requiring a compiler that one might catch errors before the JS runs on a deployed system.


> * MVC (e.g. Angular)
>   
>   * I like MVC :)
>   * Templating

name your functions "model_X", "model_Y" in plain JS and your are (for rather most intents and purposes) done.


> * Live-data binding (e.g. Angular, React)
>   
>   * Fully managed UI "loop". No need for `setTimeouts` or "callback hell" anymore.

I don't know what timeouts can be sensibly applied here and callbacks are just the necessary async continuation (Promises basically are a rename only with different syntax (similar to different mathematical notations)) but bindings give you hierarchy headaches, conglomeration concerns I don't want to have.


The white teeth advertising you propagate does not tell the truth.

This does not mean it cannot be helpful. They are tools and possibilities and given a problem they make solving easy to use them is sensible.


> I'd say they just extend the tools you have available for implementing your project. The less I have to implement, the better.

But they do not spare you any implementation of the code required to achieve a real world consumers comfort. "MVC", "DI" is a personal baggage you might like to abide by, nothing required to solve an issue.


> > needing Monstrous effort to apply for the first time... additional software on your pc, obscure storage locations, servers started on startup..
> 
> Can't really confirm that. My experience was extremely mind-blowing on what can be done with plain JS inside the node.js universe and how easy it is to apply.

Yes, plain JS can "do everything", but A can't


> Angular and React can be challenging in the beginning but the reward is really f*** huge in my opinion! Of course I'm not claiming they are here to stay forever, but `XMLHttpRequest` usage got also less popular since `fetch` is natively available if you get my point.

fetch eases XHR (just "fetch" instead of reading docs, implementing by writing many lines of code and catches), A complicates JS (need to read huge docs first, read TS doc first, decipher their intention, repel their boasting claims), A also is 50 % TS and 50 % reinventing the wheel of plain JS, locking you into them.


> > and the core of your to-be original (because if not original, why were you creating it?)
> 
> I'm mostly writing applications for to be applied not to be original. If there is a stable and relatively established solution already available for a problem which is not specific to my project (like handling a map), I'll thankfully use it. The same way I'm building on top of LNM instead of writing my own fully integrated simulator mapping, weather, traffic, flight planning and monitoring app :)

LNM and OL are comfort and solutions already, "MVC" isn't.


> > [...] from the jaw of the jail they are in
> 
> You can still keep your business logic in plain JS inside, if you're worried about those frameworks lifetimes.

does require (a bit) extra effort but yes, right.


> > ...and once it is, A has released 700 new version the readers past is useless
> 
> Yes the update cycles are fast, but this also means free out-of-the-box updates for issues you don't want to worry about. Also, it's not that bad anymore as in node.js's early days.

You have to worry about excatly about every update even in ..1 A broke compatibility (as far as I've read), before-to-last major release requiring weeks of rework (depending on project size, as I've heard) and now there is even a newer release (as I've heard). By wanting to write software to solve a real world problem and thinking about this and planning and implementing it, months if not years of net developer time, headaches and sorrow could have been avoided, but A is about hacking theoretical theories and bragging.


> > kis in every way
> 
> And that's probably the philosophical question here: Would you like to create a custom solution from scratch every time...

Like written before, A does not offer anything one would want sensibly to write and the core code has to be written. libraries like OL help, "frameworks" like A lock.

> Meaning everyone has to read-in into custom solutions everytime. Omitting the fact that they quite often lack documentation and the author has to maintain updates inside a changing environment.

This is exactly what is occuring with A ! The dependency hierachy of likely every A project is different to every other A project and so likely is the code and every A update has to be maintained and possibly lacks documentation of what they broke this time.


> In my opinion it's conventions and standards which make focusing on specific aspects of implementation work easier, leading to less specific code which translates to "simple" for me. At least I don't want to think about on how to fetch data or how to update a value in the UI every time.

Yes, of course.


> I can just look up and copy/paste the example from the corresponding frameworks docs...

Here you "fail": copy/paste you can from your previous projects as well. "Framework" already implies complex and "doc": unfortunately "good luck" finding that part documented.


> If it's for the kicks I understand the urge to create individual solutions. But not for challenges/problems which have already been solved...

This appears to be your whole derailedness of mind: no problem of your actual goal has alreay been solved (by any of the frameworks I have read about, their doc or worked with)!


> I hope you're not mad at me :) All paths lead to Rome...

I'm not. Your reply turned out well.


> If you haven't already tried out node.js, Angular and/or React please do! My guess would be, that you eventually could fall in love with at least TypeScript :)

I have (sans using React), wrt the frameworks. TypeScript I might have liked 20 years ago when I would have come directly from C++. But during working with C++ I realised it leads to attempts to engineer things which don't need enigineering, locking one-self into hierachies later not suited anymore all while costing a lot of thinking about how to express something in code rather than having everything thought through and planned before and then only implement. This is particularly true for the creatom of tools such as frameworks are which have no goal on their own. If you have planned to present a certain comfort to a user and thought through and planned all which is necessary, an OO language might make things easy, but JS is an OO language.


> Thanks for your thoughts and cheers!

Likewise :) !

**Reply by KOKAProduktion** [reply]
Hiho @u-an-i,

hmmm, you gave me something to think about... Thanks! I'm not sure if I'm with you on all the points. But for now I'll retreat and prepare my next strike :)

Just a single remark since you also seem to avoid writing out the frameworks names:
> The white teeth advertising you propagate does not tell the truth.

I didn't mean to advertise anything. I don't care whether a project uses framework A, framework B, none or something proprietary. I have to adapt to it. I'm just enjoying the ride learning from others. Sorry if I broke some code mentioning them directly...

I feel that this might go for a longer intense discussion, so maybe we can postpone it until i get this milestone done? :)

For now I'd just need confirmation from you and @albar965 if it's ok to:

- Keep the OL implementation project just about mapping and map interaction handling
    - Embeddable and configureable inside the current Web UI or any HTML/JS
- Keep the OL implementation as separate NodeJS project/repo
    - Only project build/dist output merged into LNM's `web/ol`

Thank you for your time, thoughts and cheers!

**Reply by albar965** [reply]
> For now I'd just need confirmation from you and @albar965 if it's ok to:
> 
>     * Keep the OL implementation project just about mapping and map interaction handling

I leave this up to you. 

>       * Embeddable and configureable inside the current Web UI or any HTML/JS

Yes. Maybe think about a standalone version or a GUI-less LNM option later. Neh. Just thinking.

>     * Keep the OL implementation as separate NodeJS project/repo

Yep. 
 
>       * Only project build/dist output merged into LNM's `web/ol`

Or keep as a separate project but LNM is not usable as a library. No idea. Git sub-modules to minimize dependencies perhaps? Git modules look promising.

In general it would be nice if there are no downloaded third party libraries, fonts or Google analytics links. Especially the analytics I hate.

**Reply by u-an-i** [reply]
> I feel that this might go for a longer intense discussion, so maybe we can postpone it until i get this milestone done? :)

Yes, sure! I'm not thrilled to have this conversation and am glad you took in my "observations" indulgently apparently, but I (guess I feel I) am happy to read any advantages and benefits you put forward.

**Reply by KOKAProduktion** [reply]
Finished OL integration into LNM web UI at #826

**Comment by KOKAProduktion** [comment]
## Milestone 4 : Control LNM via Web API

The idea is to extend the API established during milestone 1 by methods controlling LNM itself. This could for example provide simple functionality like changing the map theme or more complex processes like creating/modifying the flight plan. Or even add all functionaility the in-app Marble widget view offers for milestone 5.

### Requirements

- LNM Web API developed at milestone 1

### Possible Solution

- Extending the API contract by sensible methods to expose to the external client using
    - [OpenAPI](https://oai.github.io/Documentation/)
- Evaluation/Modification of LNM to allow control input coming outside the main thread
- Implementing defined methods

**Comment by KOKAProduktion** [comment]
## Milestone 5 : Replacing LNM in-app Marble with an OpenLayers web view

The idea is to replace the Marble dependency from the project to decrease build setup complexity and issues. If the solution from previous milestones proves itself reliable and satisfying it may serve as the replacement.

### Thoughts

As i understand Marble and OL are solutions serving the same purpose (Mapping) in different languages/environments. They even have similar source adapters. Also i don't see any performance-critical obstacle which would disqualify an in-app web view based solution. The only crucial thing to find out would be how good/reliable the communication between the Qt WebView JS and the QT Application works in my opinion.

### Related

- #180
- #665 

### Possible Solution

- Altering the OL implementation to retrieve map image data from Marble's current sources directly instead from LNM Marble image output
- Complementing all missing interaction logic currently provided by the in-app Marble view (if any)
- Replacing the in-app Marble Widget with a Qt WebView displaying the OL implementation

**Comment by KOKAProduktion** [comment]
# Commitment

@albar965 if you find that roadmap sketch sensible I could commit myself to milestones 1-3 on both Qt/C++ and JS parts but sadly without a specific time frame, not only for obvious reasons but also because some points still require a bit more specification.

I've managed to build LNM, thankfully to your building instructions, but coming from the web and mobile apps world my Qt/C++ commits would certainly require reviews from a Qt/C++ pro. Also I'd probably have a lot of questions during the process :) For now I'd like to say, that the sources are pleasant to read!

As for the milestones 4 and 5 i'd say, let's check again when we get there :)

Also I'm totally open to split up the work @u-an-i. If you keen on particular parts, let's elaborate a more specific plan together!

I'm looking forward to your comments. Have a nice evening!

Cheers!

**Reply by u-an-i** [reply]
thoughts:

- how can I express my awe?


observation:

- "All Good Things": status quo précis, target state aperçu, apropos linking, strategy, segmentation, conducive breakdown


fears:

- is the "Popular ... UX" not not-liked by me?


I imagine operating an OL map delivers a smooth and responsive behaviour at hardly any cost (technical, physical, financial nature), quite the contrary of the latter I imagine in fact: easy to apply, comparably lightweight and free. (i visited their website and saw your application of them.)

**Reply by KOKAProduktion** [reply]
Thank you for your kind words @u-an-i !

As for the fears, I'm not claiming that this UX is the one and only solution. Also I'm not an aviator and I surely don't know all use cases of LNM (yet), so I don't want to force anything on anyone. Neither users nor dev's :) It's just my own impression, so please express any concerns or doubts anytime!

If you don't object, I'd soon start writing the [OpenApi](https://oai.github.io/Documentation/) specification for the LNM Web API mentioned at [milestone 1](https://github.com/albar965/littlenavmap/discussions/677#discussioncomment-847061) as a base for further discussion. Or do you have already planned/started something else around that part?

Once again, thx and cheers!

**Reply by u-an-i** [reply]
i haven't planned and haven't started something around that.

look at the beginning of https://github.com/albar965/littlenavmap/discussions/661 for your consideration

**Comment by albar965** [comment]
Oi! 

I cannot say that I do not like this approach. :slightly_smiling_face: 

What I like: 
* Can be done in steps (as you described) without breaking main LNM plain and old UI.
* Getting a JSON web API as a welcome side effect.
* Makes old and ugly networking code obsolete.
* Makes Marble obsolete.
* Opens possibilities for plenty of JS add-ons/plugins as a side effect.

This will not be done in a year for sure. Just saying. :upside_down_face: 
Converting current functionality to web UI is a real challenge, I'd say.

BSD license: Not sure if this is compatible with GPL. No idea.

Any plans to use an UI framework like [Vaadin](https://vaadin.com/)?

BTW: Very professional presentation! Like a lot. :+1: 

Alex

**Reply by KOKAProduktion** [reply]
Hiho @albar965,

Cool, I'm glad you like it and thank you for the props! I'm very excited to get going!

> This will not be done in a year for sure. Just saying. 🙃
Converting current functionality to web UI is a real challenge, I'd say.

Yeah, talk is cheap... I know :) I've already had some "What have i done...?" moments browsing through the source :) but I'm still onboard and hopefully we'll be able to add usable features already on the way to the top.
As mentioned I'm still getting myself more into Qt/C++ as well as LNM itself and writing the [OpenApi](https://oai.github.io/Documentation/) specification for the future LNM Web API during the process. I hope it will serve as a base for further discussion but please give me some time.

> Opens possibilities for plenty of JS add-ons/plugins as a side effect.

Oh yes, that's a sexy one! It certainly raises and falls with the API capabilities.

> BSD license: Not sure if this is compatible with GPL. No idea.

As far as i've googled, the [BSD 2-Clause License](https://opensource.org/licenses/BSD-2-Clause) seems to be one-way compatible with GPL as long as the distributions retain its copyright notice. 

Here are some of the sources i've got this information from
- https://janelia-flyem.github.io/licenses.html
- https://opensource.stackexchange.com/a/305

Maybe we should have a closer look again before we actually start integrating OL.

> Any plans to use an UI framework like Vaadin?

I'm a fan of [Bootstrap](https://getbootstrap.com/) licensed under the [MIT License](https://getbootstrap.com/docs/5.0/about/license/) which seems also to be compatible to GPL as long as its requirements are met.

But I'm not married to it, so if you prefer any other framework I'll adjust. I'd just suggest trying to get away from server-side rendering/templating techniques to a strict REST interface only and independent, consuming client applications.

I've subscribed to the repo's notifications in order to be kept up-to-date but please keep tagging me on related items if i happen to oversee something.

Once again thanks and hear you soon hopefully!

Kosma

**Reply by albar965** [reply]
I thought about this and I'm hesitant to use the JS inside LNM for map display. Absolutely no problem with a JS plugin interface, though. Can we postpone the map display replacement to a later version? I see an interfacing nightmare, sorry. There is so much functionality on/with the map display (flight plan rubber band, modifier clicks and whatnot). Could be very difficult and work intense. 

Add things step by step.
Here an idea for a first step (maybe in next major 2.8):
* Implement JSON (REST?) interface as discussed. I already had user requests for this. Might be valuable.
* Leave the web UI as a read only interface for simplicity. Means: No flight plan editing or similar functionality.
* Port all or most map display configuration options from the view menu to the web UI. Allow people to fully configure their map display, details, theme and stuff.
* Improve information for airport and navaid display. Maybe even add a simple search.

Magic can be done later. :slightly_smiling_face: 

Lemme know what you guys think about this.

**Reply by u-an-i** [reply]
reads fine for me

**Reply by KOKAProduktion** [reply]
> Absolutely no problem with a JS plugin interface, though. 

You mean browser-based plugins using the REST API (like the Web UI) or some kind of more sophisticated JS interface enabling LNM in-app plugins/extensions?

> Can we postpone the map display replacement to a later version?

You're the boss deciding what to focus and what is merged when. No objections.

> I see an interfacing nightmare, sorry.

Absolutely, there will be blood. But i do hope that the API contract will help us a bit to oversee and control this.

> There is so much functionality on/with the map display (flight plan rubber band, modifier clicks and whatnot). Could be very difficult and work intense.

Definitely as all the objects drawing and map context interaction logic would have to be fully ported to JS. I'm with you leaving this for now.

> Leave the web UI as a read only interface for simplicity. Means: No flight plan editing or similar functionality.

As the current implementation or already OL? I'm fine with both but the OL implementation does require some modification on the Marble output to avoid visual glitches/bugs introduced by the tile-based mapping approach. (e.g. dynamic objects glitching between tiles)

The remaining points sound both reasonable and doable to me. Count me in!

Thank you and cheers!

**Reply by albar965** [reply]
> You're the boss deciding what to focus and what is merged when. No objections.

Ok. But still feel free to call me names is something is not reasonable. :upside_down_face: 

> As the current implementation or already OL? I'm fine with both but the OL implementation does require some modification on the Marble output to avoid visual glitches/bugs introduced by the tile-based mapping approach. (e.g. dynamic objects glitching between tiles)

OL would be cool and what you have as a prototype is very fluid. I leave this up to you. 

1. No OL. Marble delivers full images as is right now. Inefficient but least effort.
1. Use OL. Slight modifications in Marble to leave out the copyright labels. Not very efficient to download tiles, render a full map and then use this as a tile.
1. Use OL and Marble HttpDownloadManager or another class to get tiles and pass them as they are to the web UI. Most efficient approach. LNM then acts more as a tile proxy. Needs changes and some research on Marble internals.

BTW: I'd limit the projection in the web UI to Mercator since the tiles use the same projection. 

You can take your time since 2.8 won't be release soon, i.e. not this year. I'd rather do minor updates which fix bugs and a few cherry picked changes.

Alex
